cmd/fleet/vuln_process.go

```diff
@@ -0,0 +1,161 @@
+	"github.com/fleetdm/fleet/v4/server/datastore/mysql"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	kitlog "github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/spf13/cobra"
+)
+
+var (
+	dev               bool
+	devLicense        bool
+	devExpiredLicense bool
+	lockDuration      time.Duration
+)
+
+func createVulnProcessingCmd(configManager config.Manager) *cobra.Command {
+	vulnProcessingCmd := &cobra.Command{
+		Use:   "vuln_processing",
+		Short: "Run the vulnerability processing features of Fleet",
+		Long: `The vuln_processing command is intended for advanced configurations that want to externally manage 
+vulnerability processing. By default the Fleet server command internally manages vulnerability processing via scheduled
+'cron' style jobs, but setting 'vulnerabilities.disable_schedule=true' or 'FLEET_VULNERABILITIES_DISABLE_SCHEDULE=true' 
+will disable it on the server allowing the user configure their own 'cron' mechanism`,
+		Run: func(cmd *cobra.Command, args []string) {
+			cfg := configManager.LoadConfig()
+			if dev {
+				applyDevFlags(&cfg)
+			}
+
+			licenseInfo, err := initLicense(cfg, devLicense, devExpiredLicense)
+			if err != nil {
+				initFatal(
+					err,
+					"failed to load license - for help use https://fleetdm.com/contact",
+				)
+			}
+
+			if licenseInfo != nil && licenseInfo.IsPremium() && licenseInfo.IsExpired() {
+				fleet.WriteExpiredLicenseBanner(os.Stderr)
+			}
+
+			ds, err := mysql.New(cfg.Mysql, clock.C)
+			if err != nil {
+				initFatal(err, "creating db connection")
+			}
+
+			// we need to ensure this command isn't running with an out-of-date database
+			status, err := ds.MigrationStatus(cmd.Context())
+			if err != nil {
+				initFatal(err, "retrieving migration status")
+			}
+
+			var migrationError error
+			switch status.StatusCode {
+			case fleet.AllMigrationsCompleted:
+				// only continue if db is considered up-to-date
+			case fleet.NoMigrationsCompleted:
+				migrationError = errors.New("no migrations completed")
+			case fleet.SomeMigrationsCompleted:
+				migrationError = errors.New("partial migrations completed")
+			case fleet.UnknownMigrations:
+				migrationError = errors.New("database migrations incompatible with current version")
+			}
+			if migrationError != nil {
+				initFatal(migrationError, "refusing to continue processing vulnerabilities, database out of sync")
+			}
+
+			ctx := context.Background()
+			ctx = license.NewContext(ctx, licenseInfo)
+			instanceID, err := server.GenerateRandomText(64)
+			if err != nil {
+				initFatal(errors.New("error generating random instance identifier"), "")
+			}
+			// using the same lock name as the cron scheduled version of vuln processing, that way if we fail to obtain the lock
+			// it's most likely due to vulnerabilities.disable_schedule=false but still trying to run external vuln processing command
+			lock, err := ds.Lock(ctx, "cron_vulnerabilities", instanceID, lockDuration)
```

Comment:
Curious to know why not use the `schedule` package.

Reply:
eh mostly just because it didn't feel right. The schedule package has a lot of baggage around...well scheduling. Sure it does some locking, and extending of said locks, but I didn't see an easy way to grab the things I need, aka locking, without taking a whole lot of things I didn't.

I suppose I could have just created a `schedule.New()` with effectively no interval, and then call `Trigger()` on it, but `Trigger()` simply pushes a message into a channel and returns, there is effectively no way for me to know when the thing is done.

Typically in these kinds of scenarios I'd reach for something simple like a [redis distributed lock](https://redis.io/docs/manual/patterns/distributed-locks/) of which there are a handful of decent Go implementations, but I didn't want to introduce a secondary locking mechanism without further discussion.

If you see an easy path to utilize schedule, I'm happy to oblige.

Reply:
> I suppose I could have just created a schedule.New() with effectively no interval, and then call Trigger() on it, but Trigger() simply pushes a message into a channel and returns, there is effectively no way for me to know when the thing is done.

Good feedback :)

> Typically in these kinds of scenarios I'd reach for something simple like a [redis distributed lock](https://redis.io/docs/manual/patterns/distributed-locks/) of which there are a handful of decent Go implementations, but I didn't want to introduce a secondary locking mechanism without further discussion.

I believe this is a-ok, the only reason I wanted to use the `schedule` package is for the capability it has to "extend the lock" if the task is taking longer than the configured periodicity of the cron.
E.g. somebody configures `fleet vuln_processing` to run every 1 hour, what happens if vulnerability processing takes longer than 1 hour? (an expired lock is considered "unlocked") Maybe to be on the safe side we could call `initFatal`  if it takes longer than the lock duration?

Reply:
How do you think this would look? Something like:

```go
go func() {
  <- time.Tick(lockDuration)
  cancel()
  initFatal(errors.New("processing time exceeds lock duration"), "")
}()

go func() {
  //scanVulnerabilities(ctx, etc, etc)
}()
```

Reply:
But for a first version I believe this is a-ok. Let's give this a go first and get some real-world feedback.

Reply:
Ya, thinking it twice am not sure about killing the process when it could be writing to the DB (maybe it will leave the tables in an invalid state?) So am ok with tackling this later if really need be.